crate/server/src/database/redis/mod.rs

```diff
@@ -1,130 +1,190 @@
-use argon2::Argon2;
-use async_trait::async_trait;
-use cloudproof::reexport::crypto_core::{reexport::rand_core::SeedableRng, FixedSizeCBytes};
-use std::{
-    collections::HashSet,
-    sync::{Arc, Mutex},
-};
+use std::collections::{HashMap, HashSet};
 
-use cloudproof::reexport::crypto_core::{kdf256, Aes256Gcm, CsRng, Instantiable, SymmetricKey};
+use async_trait::async_trait;
 use cloudproof_findex::{
-    implementations::redis::{FindexRedis, FindexRedisError, RemovedLocationsFinder},
-    parameters::MASTER_KEY_LENGTH,
-    Label, Location,
+    db_interfaces::{
+        redis::{build_key, FindexTable, TABLE_PREFIX_LENGTH},
+        rest::UpsertData,
+    },
+    reexport::cosmian_findex::{
+        CoreError, EncryptedValue, Token, TokenToEncryptedValueMap, TokenWithEncryptedValueList,
+        Tokens, ENTRY_LENGTH, LINK_LENGTH,
+    },
 };
-use redis::aio::ConnectionManager;
-
-use crate::{error::FindexServerError, result::FResult, secret::Secret};
+use redis::{aio::ConnectionManager, pipe, AsyncCommands, Script};
+use tracing::{instrument, trace};
 
 use super::Database;
-
-pub(crate) const REDIS_WITH_FINDEX_MASTER_KEY_LENGTH: usize = 32;
-pub(crate) const REDIS_WITH_FINDEX_MASTER_KEY_DERIVATION_SALT: &[u8; 16] = b"rediswithfindex_";
-pub(crate) const REDIS_WITH_FINDEX_MASTER_FINDEX_KEY_DERIVATION_SALT: &[u8; 6] = b"findex";
-pub(crate) const REDIS_WITH_FINDEX_MASTER_DB_KEY_DERIVATION_SALT: &[u8; 2] = b"db";
-
-pub(crate) const DB_KEY_LENGTH: usize = 32;
+use crate::error::{result::FResult, server::FindexServerError};
+
+// TODO(manu): miss the clear_database parameter
+// TODO(manu): move secret to client crate
+
+/// The conditional upsert script used to only update a table if the
+/// indexed value matches ARGV[2]. When the value does not match, the
+/// indexed value is returned.
+const CONDITIONAL_UPSERT_SCRIPT: &str = r"
+        local value=redis.call('GET',ARGV[1])
+        if((value==false) or (not(value == false) and (ARGV[2] == value))) then
```

Comment:
If Lua's type system makes any sense, the first check of the right condition of the `or` should be redundant with the left condition.
```suggestion
        if ((value==false) or (ARGV[2] == value)) then
```

Reply:
right